impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java

```diff
@@ -0,0 +1,241 @@
+
+      streamListener = new StreamExecutionListener( buffer );
+
+      serviceListeners.put( cacheId, streamListener );
+    }
+
+    if ( !isRunning.get() ) {
+      isRunning.compareAndSet( false, true );
+      startService();
+    }
+
+    return streamListener;
+  }
+
+  /**
+   * Generates the cache key for a given query with a specific size and rate.
+   *
+   * @param query The query.
+   * @param size The query window size.
+   * @param millis The query window rate.
+   * @return The cache key for the query.
+   */
+  private String getCacheKey( String query, int size, long millis ) {
+    return query.concat( String.valueOf( size ) ).concat( String.valueOf( millis ) );
+  }
+
+  /**
+   * Starts the Service transformation and its row event listener.
+   */
+  private void startService() {
+    try {
+      lastCacheCleanupMillis = new Date().getTime();
+      serviceTrans.startThreads();
+
+      StepInterface serviceStep = serviceTrans.findRunThread( serviceStepName );
+      if ( serviceStep == null ) {
+        throw Throwables.propagate( new KettleException( "Service step is not accessible" ) );
+      }
+      serviceStep.addRowListener( new RowAdapter() {
+        /**
+         * Listener for the service transformation output rows.
+         * If the stepStream has any valid registered listeners it writes the output row to the stepStream,
+         * otherwise stops the service transformation and kills its running thread.
+         *
+         * @param rowMeta The metadata of the written row.
+         * @param row The data of the written row.
+         * @throws KettleStepException
+         */
+        @Override
+        public void rowWrittenEvent( RowMetaInterface rowMeta, Object[] row ) throws KettleStepException {
+          // Simply pass along the row to the other transformation (to the Injector step)
+          LogChannelInterface log = serviceTrans.getLogChannel();
+
+          try {
+            if ( log.isRowLevel() ) {
+              log.logRowlevel( DataServiceConstants.PASSING_ALONG_ROW + rowMeta.getString( row ) );
+            }
+          } catch ( KettleValueException e ) {
+            // Ignore error
+          }
+
+          RowMetaAndData rowData = new RowMetaAndData();
+          rowData.setRowMeta( rowMeta );
+          rowData.setData( row );
+
+          serviceCacheCleanup();
+
+          if ( serviceListeners.size() > 0 ) {
+            stepStream.add( rowData );
+          } else {
+            isRunning.compareAndSet( true, false );
+            serviceTrans.stopAll();
+            serviceTrans.waitUntilFinished();
+            stepStream = null;
+            log.logDebug( DataServiceConstants.STREAMING_TRANSFORMATION_STOPPED );
+          }
+        }
+      } );
+    } catch ( KettleException e ) {
+      throw Throwables.propagate( e );
+    }
+  }
+
+  /**
+   * Clean um service listener cache.
+   */
+  private void serviceCacheCleanup() {
+    long currentTime = new Date().getTime();
```

Comment:
Consider using `System.currentTimeMillis` instead, since it has better performance (not much, from `Date().getTime()` I reckon), but it will save on GC execution, by skipping the creation of a new object, that fills memory and then has to be cleaned up.

Same on [StreamingServiceTransExecutor](https://github.com/pentaho/pdi-dataservice-server-plugin/pull/279/files#diff-b6aad85ffbd8f09c0d85e7ff8295f8bf), but more relevant here, since this is called for each row.